transcom/mymove
# MB-16387 Fix submit ppm docs endpoint authz gap

## [Jira ticket](https://dp3.atlassian.net/browse/MB-16387)

## Summary

### Background

We have a [discovery doc that outlines the API endpoints that have gaps in authz checks, as well as a plan to address them](https://dp3.atlassian.net/wiki/spaces/MT/pages/2115043329/Discovery+-+Customer+Internal+API+Authorization+Issues) 🔒. Part of what we decided was that we would include user information as part of the query when possible in order to let the DB handle the search and avoid us even loading the data (vs loading the data and then checking if the service member matches). This also easily sets us up to return a 404 when the user doesn't match, which is good because this way we won't even confirm that the ID they tried using is associated with anything (in this case we use order ID).

### Changes

<details>
<summary>Main Changes (i.e. things that explicitly satisfy AC)</summary>

* [feat (service): update GetPPMShipment to account for customer authz](https://github.com/transcom/mymove/pull/11209/commits/eb487bea9b0f10a361c470e01bacda6e7899b9f3)
    * Should work the way it was before if the request is coming from the prime API or the office app. Only adds authz check if it is from the customer app. If it is a customer request, we ensure the service member in the session matches the one that the shipment belongs to (via orders association).
    * Updated `moving_expense_creator` and `progear_weight_ticket_creator` to call `GetPPMShipment` differently now that it has the logic to use the logged in customer info as appropriate. Also updated `moving_expense_creator` to pass the errors from `GetPPMShipment` along rather than turning them into server errors.
* [fix (api): ensure customer authz are checked for ppm closeout submission](https://github.com/transcom/mymove/pull/11209/commits/b41ca60e82c68d37a67c79f7517e6438fcf930e0)
    * We accomplish this by using the `GetPPMShipment` service instead of `FetchPPMShipment` in the `SubmitNewCustomerCloseOut` service. This also gives us the ability to mock out the dependency, which wasn't possible before.

</details>

<details>
<summary>Other Changes</summary>

* [feat (test): Update a PPM shipment factory to allow customizations](https://github.com/transcom/mymove/pull/11209/commits/030eb9af41123e8eb654c1293162fdac54b800b1)
    * This was needed for the integration tests w/routing & middleware since they need an active user, which we enable via customizations.

</details>

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.

### Setup to Run the Code

- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/about/application-setup/milmove-local-client/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/about/development/testing)

### How to test

1. Start server and customer client.
2. Log in as a service member that is ready to submit their PPM closeout documentation (e.g. `weightTicketPPM@ppm.approved`)
    1. Get the ID of their PPM shipment from the network tab. Look at the `mto_shipments` response info to see it. The tree of data should be `[0].ppmShipment.id` 
    2. ![image](https://github.com/transcom/mymove/assets/35938642/4771fa6e-0495-44ca-af28-0d59996b51ad)
3. Log out.
4. Log in as another service member to get their cookie information for postman. Load into postman based on however you set up postman.
    1. [Docs for setting up postman](https://transcom.github.io/mymove-docs/docs/tools/postman/setting-up-postman)
5. Make a `POST` request in postman to `/ppm-shipments/ :ppmShipmentId/submit-ppm-shipment-documentation` 
    1. I made the body a raw JSON request with this:
        1. ```json
           {
             "certification_text": "I agree to all the terms and conditions, including changes into perpetuity!",
             "signature": "Easiest Customer",
             "date": "2023-08-15"
           }
           ```
    1. You can swap out the `:ppmShipmentId` in the path with the ID you got earlier, or use the `Params` tab to set the `Path Variable` value for `ppmShipmentId`.
    1. Since this is going to be a patch request, you'll need to [set the `X-CSRF-TOKEN` header to be the `masked_gorilla_csrf` cookie value](https://transcom.github.io/mymove-docs/docs/tools/postman/intercepting-cookies-for-postman#form-a-working-get-and-patch-request-in-postman).
7. Verify you get a 404 response.
    1. ![image](https://github.com/transcom/mymove/assets/35938642/b4ccb821-6844-496c-aa60-75ee6dc3e76b)
    1. If you try the above steps on the main branch, you'll get a 200 instead indicating the closeout docs were submitted successfully.

### Backend
